### PR TITLE
Adds stable JSON type support for the Cpanel::JSON::XS encoder

### PR DESCRIPTION
**Pull request by pali**
This patch series provides stable JSON type support for the Cpanel::JSON::XS encoder which doesn't depend on any internal perl scalar flags or characteristics.

In most cases perl structures passed to encode_json come from other functions or from other modules and caller of Cpanel::JSON::XS module does not have control of internals or they are subject of change. So it is not possible to "force" type as described in the simple scalars section.

Example:
```
use Cpanel::JSON::XS;
use Cpanel::JSON::XS::Type

# ExternalModule::external_function is not under our control
my $perl_struct = ExternalModule::external_function($some_input_data);
my $type_spec = json_type_hashof(JSON_TYPE_INT);
my $json_string = encode_json($perl_struct, $type_spec);
# $json_string now contains only JSON objects with JSON numeric values without fractional part (integers)
```

Function encode_json takes perl structure as its input and optionally also json type specification in the second parameter.

If the specification is not provided (or is undef) internal perl scalar flags are used for choosing JSON type. Internal flags can be changed by perl itself, also by external modules which means that types in resulting JSON string aren't stable.

This patch series contains also other changes, specially fixing processing magic scalars and making Cpanel::JSON::XS::true() and Cpanel::JSON::XS::false() inline-able functions.

### CODE REVIEW

**Comment by pali** [comment]
AppVeyor tests are failing because they are unable to download StrawberryPerl, looks like strawberryperl.com is down:
```
ERROR: The remote file either doesn't exist, is unauthorized, or is forbidden for url 'http://strawberryperl.com/download/5.26.0.1/strawberry-perl-5.26.0.1-64bit.msi'. Exception calling "GetResponse" with "0" argument(s): "The remote name could not be resolved: 'strawberryperl.com'"
```

**Comment by rurban** [comment]
I would rather look what other typed json modules offer, such as eg TypeScript or nim or elm, and implement that. Perl can do types, cperl even more, so it makes no sense to deviate from that. Special type hooks are also easy to add creating a custom result from the read json type.

**Comment by pali** [comment]
> Perl can do types, cperl even more, so it makes no sense to deviate from that

Primary use of Cpanel::JSON::XS is in perl, not cperl and I think more people want to use it in perl.

Perl itself does not have strict types and what it have is ambiguous. First look at perlnumber:
http://perldoc.perl.org/perlnumber.html

If you want to define types by internal and private SvFLAGS then it is wrong. Look at following number:
```
$ perl -MDevel::Peek -e 'my $num = "10e10"; Dump $num'
SV = PV(0x5590884e2b20) at 0x559088509cb0
  REFCNT = 1
  FLAGS = (POK,IsCOW,pPOK)
  PV = 0x55908850e320 "10e10"\0
  CUR = 7
  LEN = 10
  COW_REFCNT = 1
```
Your Cpanel::JSON::XS thinks it is a string, but that is not truth, it is mean to be a number in this context.

Or look at another example:
```
$ perl -MDevel::Peek -e 'my $num = 10; if (length($num) > 10) { $num = 9999999999 } Dump $num;'
SV = PVIV(0x55c84ec4dbb0) at 0x55c84ec51cd0
  REFCNT = 1
  FLAGS = (IOK,POK,pIOK,pPOK)
  IV = 10
  PV = 0x55c84ec426b0 "10"\0
  CUR = 2
  LEN = 10
```
Here caller needs to ensure that number would not have more then 10 digits. This is perl way how to do it. But Cpanel::JSON::XS misuse SvFLAGS for types and thinks that it both string and integer at same time. But it is a number, not string.
___

And in more cases output structure which needs to be enabled to JSON comes from external module or from external source (e.g. database) which means that scalars can contain any SvFLAGS.

Also perlnumber specify 7 equivalent ways how to write number in perl, which lot of modules do as it is perl way. But Cpanel::JSON::XS does not see all those ways to be numbers. And here is a problem. Perl scalar is mean to be without type and you should not use SvFLAGS for trying to guess what is the type. JSON format is strictly typed and once if you need to have number, then you should be able to tell JSON encoder that you want number. Plus Cpanel::JSON::XS uses *private* SvFLAGS...

In cperl it is different situation, but I'm primary interested in perl.

**Comment by rurban** [comment]
I'm talking about cperl coretypes, matching perl6 types, which can be used with perl5 also in lexical decls. Not signatures, that's cperl and perl6 only.

`'my num $num = "10e10"; Dump $num'`

**Comment by pali** [comment]
That is truth, but does not solve problem when whole structure comes from external module which do not care about types. Specially PostgreSQL or text file parser which returns everything as string (because for parsing text files are used regexes).

As wrote at beginning motivation is ability to produce JSON strings with stable types generated from structured returned by external modules and functions which are not under control of JSON encoder caller. Therefore it is not needed to use types there.

**Comment by rurban** [comment]
If we are going to add an optional Typed JSON module, we'll have to use the established interfaces.

Which are perl6/cperl on one hand, and the various typed json modules in other languages on the other hand. perl6/cperl can do typed classes (structs), arrays and hashes, perl5 only scalars. So that's a dilemma. Will think about it. As I said, a new typed callback sounds best. This is how some other langs do it. Not via a template for sure, then rather proper types via destructuring bind/match. But not even perl6 has that yet.

**Comment by pali** [comment]
Yea, it was hard decision how to handle this problem... but in case when all data and whole complicated perl structure which need to be encoded comes from external modules or external database (e.g. PostgreSQL or MySQL) which do not support types either in perl5 modules or in transfer protocol (MySQL for non-prepared statements supports only STRINGs) then there is the only option to define "template" for encoder. Basically here we needs to deal with "unknown" or "unspecified" types from external world and we need to have stable and specific type in JSON on output which would be independent on used database backend or other cpan modules...

Currently we chose and implemented this "template" based option as it fully covers our real use cases in GoodData. We also considered other options like calling `int()` or `""` operators, but such thing is very fragile, slow and not robust like current solution where type is 100% stable independently of external cpan modules or database from which data comes from.

Basically for real use-cases and big systems there is no better option...

**Comment by rurban** [comment]
I'll work on this, but expect something earliest in September.

**Comment by pali** [comment]
I have not fully understood your last comment. Are you going to propose different encoder API for enforcing JSON types from arbitrary perl structure?

And what do you think about current API implementation? Would you accept it, or needs to be slightly reworked?

**Comment by rurban** [comment]
I have no time currently to look at it. 2 major releases, TPF talk, OO system fixes, YAML security, ...

**Comment by pali** [comment]
All tests now passed, inlining of the true() and false() is fixed.

**Comment by pali** [comment]
Hi Reini! Any comments or review on this PR?

**Comment by rurban** [comment]
Not yet, still busy with safeclib.

**Comment by pali** [comment]
Any news? I would really like to know if this pull request can be merged, or if it is needed to rework something...

**Comment by pali** [comment]
@rurban Do you have a time to look at this pull request?

**Comment by rurban** [comment]
I think it looks pretty good so far. see the pr94-type branch.
I'll improve the pod a bit. 
My biggest problem is with the removal of the magic hooks. I'll have to think about that. Do yo have any test cases for these?

**Comment by pali** [comment]
I do not have a test for magic changes, but I removed duplicate magic calls as there are explicit `SvGETMAGIC (sv)` or similar calls. Basically it is not a good idea to call get magic more times in XS function for one variable.

**Comment by pali** [comment]
About Scalar::Utils and Perl 5.6.2... it is working (again), so you do not have to disable it in Perl 5.6.2 tests...
https://github.com/Dual-Life/Scalar-List-Utils/pull/55
http://matrix.cpantesters.org/?dist=Scalar-List-Utils+1.49

**Comment by pali** [comment]
Anyway, feel free to improve/modify POD documentation if it is needed. I'm fine with changes in pr94-type branch.

**Comment by rurban** [comment]
Do you have time for benchmarks? That's the last bit missing, how much
slower the general case is (est <1%) and the typed case.

Am 25.01.2018 15:58 schrieb "pali" <notifications@github.com>:

Anyway, feel free to improve/modify POD documentation if it is needed. I'm
fine with changes in pr94-type branch.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/rurban/Cpanel-JSON-XS/pull/94#issuecomment-360490927>,
or mute the thread

**Comment by pali** [comment]
Our team did some benchmarking and result is: The Cpanel::JSON::XS is considerably faster (approx. 15-20%) than the older JSON::XS. There is no appreciable difference (approx. 1%) between this patched and the upstream versions of Cpanel::JSON::XS (when type argument is not passed).

```
RHEL6:
JSON::XS version 2.27: 66.686845 second(s)
Cpanel::JSON::XS version 3.023 + type patches: 54.532425 second(s)
Cpanel::JSON::XS version 3.029 (upstream variant): 55.787079 second(s)

RHEL7:
JSON::XS version 3.01: 37.629924 second(s)
Cpanel::JSON::XS version 3.023 + type patches: 32.66306 second(s)
Cpanel::JSON::XS version 3.029 (upstream variant): 32.967041 second(s)
```

And next we compared Cpanel::JSON::XS with these patches and specifying type via optional second argument against old JSON::XS. Time measurement in seconds is below:
```
  type argument       old JSON::XS
0.221262931823730, 0.204293966293335
0.235275983810425, 0.218708992004395
0.226047992706299, 0.212330102920532
0.256052017211914, 0.233234882354736
0.241660118103027, 0.250401020050049
0.250277996063232, 0.236065149307251
0.254600048065186, 0.232470035552979
0.254650831222534, 0.239913940429688
0.232930898666382, 0.220671176910400
0.232760906219482, 0.211783885955811
0.228220939636230, 0.218284130096436
0.247806072235107, 0.235421895980835
0.243855953216553, 0.221603155136108
0.240880012512207, 0.219773054122925
0.239836931228638, 0.224635124206543
0.263254165649414, 0.256259918212891
0.242079973220825, 0.229122161865234
0.244446992874146, 0.247294902801514
0.222602844238281, 0.207175970077515
```

So when specifying optional type argument, encoder is slightly slower, but it is not recognizable.

**Comment by rurban** [comment]
Preparing the 4.00 release now.

**Comment by rurban** [comment]
Done with 3.99_01, will be in 4.00
